### PR TITLE
Add preserveEntrypoint option to splitChunks cacheGroups (rebased)

### LINKED ISSUE #7230: Webpack 4 splitChunks causes reused chunks to not be executed
# Bug report

I am working in a complicated app that has a bunch of bundles on every page. Most of these are "global" bundles and appear on every page. So, I have configured webpack to prevent any modules that appear in these bundles from appearing in any other bundle via splitChunks. More context on this here: https://stackoverflow.com/questions/49163684/how-to-configure-webpack-4-to-prevent-chunks-from-list-of-entry-points-appearing

**What is the current behavior?**

It seems that when I configure splitChunks in this way, it causes the global bundles to not be executed.

**If the current behavior is a bug, please provide the steps to reproduce.**

Here's a branch on a repo that reproduces this issue: https://github.com/lencioni/webpack-splitchunks-playground/tree/splitchunks-execution-problem

In this repro case, it currently logs the following to the console:

```
core module
core module b
non-core module
non-core b2
non-core module b
```

In the debugging I've done, it seems that this splitChunks configuration causes the `executeModules` part of the bundle that the runtime chunk is looking for to be `undefined`.

**What is the expected behavior?**

The code in the "global" bundles should be executed, which should cause the console log to look more like this:

```
core module
core bundle
core module b
core bundle b
core bundle c
non-core module
non-core b2
non-core module b
```

**Other relevant information:**
webpack version:  4.8.1
Node.js version: 8.9.1
Operating System: Mac
Additional tools:

### PR DESCRIPTION
**Pull request by JanJakes**
This is just a fork of #7242 rebased on master since the original pull request was not rebased.

Fixes #7230, closes #7242.

### CODE REVIEW

File: test/__snapshots__/StatsTestCases.test.js.snap [hunk 1]
```diff
@@ -3200,6 +3200,14 @@ Child enforce-min-size:
         > ./ main"
 `;
 
+exports[`StatsTestCases should print correct stats for split-chunks-issue-7230 1`] = `
+"         Asset      Size  Chunks             Chunk Names
+core.bundle.js  6.59 KiB       0  [emitted]  core
+main.bundle.js  6.83 KiB       1  [emitted]  main
+Entrypoint core = core.bundle.js
+Entrypoint main = core.bundle.js main.bundle.js"
```
Comment (sokra) [inline test/__snapshots__/StatsTestCases.test.js.snap]: This is a non-working bundle.

both core and main include the runtime. Using both will throw an error about missing modules at runtime. main can't use the modules in core.

Comment (JanJakes) [reply test/__snapshots__/StatsTestCases.test.js.snap]: I thought the original pull request needed just a rebase. I'll look into it. Do you have any suggestions what needs to be done?

Also - I'm not sure why there was such a complex test function https://github.com/webpack/webpack/pull/8614/files#diff-741d374868f6de786477bdcfbaf6fa81R32, the test can be simpler, right?

And looking at it some more, maybe `Preserve existing chunk as an entrypoint when it has matching modules` is not a very clear description, or is it? I'd suggest something like `Preserve entrypoint when it has same name name as chunk`?

Comment (JanJakes) [reply test/__snapshots__/StatsTestCases.test.js.snap]: @sokra Fixed. I fixed and rewrote the test to be simpler and more readable & improved the description.

**Comment by JanJakes** [comment]
@sokra Does it look better now or does it need any more work?

**Comment by JanJakes** [comment]
Rebased on master.

**Comment by JanJakes** [comment]
@sokra Could you (or somebody) please provide any more feedback here? I worked on this since rebase was [requested](https://github.com/webpack/webpack/pull/7242#issuecomment-444135811) on the original PR and I'd like to push it forward so it doesn't die abandoned.

**Comment by egorvoronov** [comment]
Faced the same issue. Any input why we couldn't put our common code to the entrypoint from lazy chunks, for example?

**Comment by ianks** [comment]
This functionality is critical for us, any chance for a review @sokra

**Comment by evilebottnawi** [comment]
@JanJakes can you rebase?

**Comment by JanJakes** [comment]
@evilebottnawi Rebased but look at the failures:
```
ERROR in SplitChunksPlugin
Cache group "core" conflicts with existing chunk.
Both have the same name "core".
Use a different name for the cache group.
HINT: You can omit "name" to automatically create a name.
BREAKING CHANGE: webpack < 5 used to allow to use the entrypoint as splitChunk. This is no longer allowed. Remove this entrypoint and add modules to cache group's 'test' instead. If you need modules to be evaluated on startup, add them to the existing entrypoints (make them arrays). See migration guide of more info.
```

I'm not familiar with changes for 5.0. Does this mean this is no longer relevant?

**Comment by niksy** [comment]
What’s the status of this issue? Is there any alternative for this in Webpack 5?

**Comment by alexander-akait** [comment]
@JanJakes Can you rebase?

**Comment by JanJakes** [comment]
@alexander-akait Not sure, after the last rebase almost two years ago I run into some errors and was asking whether this is still relevant (see https://github.com/webpack/webpack/pull/8614#issuecomment-564942252) for Webpack 5.

Maybe we can simply close this issue?

**Comment by alexander-akait** [comment]
It means we break something...

**Comment by JanJakes** [comment]
I'm closing this as I have received no feedback from the Webpack team, I'm not using it anymore, and the original issue was [rejected](https://github.com/webpack/webpack/pull/7242#issuecomment-603906164) anyway. If this is still relevant for anyone, feel free to use this branch as a base for another pull request.

**Pull request closed without merging**